dispatcher/producers/scheduled.py

```diff
@@ -1,49 +1,65 @@
 import asyncio
 import logging
-from typing import Union
+import time
+from typing import Optional, Union
 
+from ..protocols import DispatcherMain
+from ..service.next_wakeup_runner import HasWakeup, NextWakeupRunner
 from .base import BaseProducer
 
 logger = logging.getLogger(__name__)
 
 
+class ScheduleEntry(HasWakeup):
+    def __init__(self, period: float, start_time: float, body: dict) -> None:
+        self.period = period
+        self.last_ran = start_time
+        self.body = body
+
+    def mark_run(self) -> None:
+        self.last_ran = time.monotonic()
+
+    def next_wakeup(self) -> Optional[float]:
+        return self.last_ran + self.period
+
+
 class ScheduledProducer(BaseProducer):
-    def __init__(self, task_schedule: dict[str, dict[str, Union[int, str]]]):
+    def __init__(self, task_schedule: dict[str, dict[str, Union[int, str]]]) -> None:
         self.task_schedule = task_schedule
-        self.scheduled_tasks: list[asyncio.Task] = []
+        self.schedule_entries: set[ScheduleEntry] = set()
+        self.dispatcher: Optional[DispatcherMain] = None
+        self.schedule_runner = NextWakeupRunner(self.schedule_entries, self.trigger_schedule)
         super().__init__()
 
-    async def start_producing(self, dispatcher) -> None:
+    async def trigger_schedule(self, entry: ScheduleEntry) -> None:
+        entry.mark_run()
+        self.produced_count += 1
+        message = entry.body.copy()
+        message['uuid'] = f'sch-{self.produced_count}'
```

Comment:
I'm missing context of what this is used for, but is a predictable string here good enough as a uuid?

Reply:
This isn't the first time someone pointed this out. I filed https://github.com/ansible/dispatcherd/issues/96 to fix that.